libsrc/cbm/tgi/cbm-lo.s

```diff
@@ -0,0 +1,456 @@
+
+; ------------------------------------------------------------------------
+; UNINSTALL routine. Is called before the driver is removed from memory. May
+; clean up anything done by INSTALL but is probably empty most of the time.
+;
+; Must set an error code: NO
+;
+
+UNINSTALL:
+        rts
+
+; ------------------------------------------------------------------------
+; INIT: Changes an already installed device from text mode to graphics
+; mode.
+; Note that INIT/DONE may be called multiple times while the driver
+; is loaded, while INSTALL is only called once, so any code that is needed
+; to initializes variables and so on must go here. Setting palette and
+; clearing the screen is not needed because this is called by the graphics
+; kernel later.
+; The graphics kernel will never call INIT when a graphics mode is already
+; active, so there is no need to protect against that.
+;
+; Must set an error code: YES
+;
+
+INIT:
+
+; Done, reset the error code
+
+        lda     #TGI_ERR_OK
+        sta     ERROR
+;       rts
+
+; ------------------------------------------------------------------------
+; DONE: Will be called to switch the graphics device back into text mode.
+; The graphics kernel will never call DONE when no graphics mode is active,
+; so there is no need to protect against that.
+;
+; Must set an error code: NO
+;
+
+;DONE:
+
+; ------------------------------------------------------------------------
+; GETERROR: Return the error code in A and clear it.
+
+GETERROR:
+        ldx     #TGI_ERR_OK
+        lda     ERROR
+        stx     ERROR
+        rts
+
+; ------------------------------------------------------------------------
+; CONTROL: Platform/driver specific entry point.
+;
+; Must set an error code: YES
+;
+
+CONTROL:
+        lda     #TGI_ERR_INV_FUNC
+        sta     ERROR
+        rts
+
+; ------------------------------------------------------------------------
+; CLEAR: Clears the screen.
+;
+; Must set an error code: NO
+;
+
+CLEAR:
+        lda     #0
+        tay
+@L1:    sta     COPY+$0000,y
+        sta     COPY+$0100,y
+        sta     COPY+$0200,y
+        sta     COPY+$0300,y
+        sta     COPY+$0400,y
+        sta     COPY+$0500,y
+        sta     COPY+$0600,y
+        sta     COPY+$0652,y
+        iny
+        bne     @L1
+DONE:
+        lda     #147
+        jmp     BSOUT
+
+; ------------------------------------------------------------------------
+; SETCOLOR: Set the drawing color (in A). The new color is already checked
+; to be in a valid range (0..maxcolor-1).
+;
+; Must set an error code: NO (will only be called if color ok)
+;
+
+SETCOLOR:
+        sta     COLOR
+;       rts
+
+; ------------------------------------------------------------------------
+; SETVIEWPAGE: Set the visible page. Called with the new page in A (0..n).
+; The page number is already checked to be valid by the graphics kernel.
+;
+; Must set an error code: NO (will only be called if page ok)
+;
+
+SETVIEWPAGE:
+;       rts
```

Comment:
My personal opinion was/is that we deliberately keep all entry points to drivers unique. However, recently it came into fashion to fold them together. But that change of style wasn't done throughout the code base, so it's now less consistent then before :-(

Let's see what rolling the dice will result in this driver ;-))

Reply:
Did you see my optimisation for DONE? I didn't want to repeat the code.